src/java.desktop/macosx/native/libawt_lwawt/java2d/metal/MTLGraphicsConfig.m

```diff
@@ -0,0 +1,268 @@
+    (JNIEnv *env, jclass mtlgc)
+{
+    FILE *f = popen("/usr/sbin/system_profiler SPDisplaysDataType", "r");
+    bool metalSupported = JNI_FALSE;
+    while (getc(f) != EOF)
+    {
+        char str[60];
+
+        if (fgets(str, 60, f) != NULL) {
+            // Check for string
+            // "Metal:  Supported, feature set macOS GPUFamily1 v4"
+            if (strstr(str, "Metal") != NULL) {
+                //puts(str);
+                metalSupported = JNI_TRUE;
+                break;
+            }
+        }
+    }
+    pclose(f);
+
+#ifdef METAL_DEBUG
+    if (!metalSupported) {
+        fprintf(stderr, "Metal support not present\n");
+    } else {
+        fprintf(stderr, "Metal support is present\n");
+    }
+#endif
+
+    J2dRlsTraceLn1(J2D_TRACE_INFO, "MTLGraphicsConfig_isMetalFrameworkAvailable : %d", metalSupported);
+
+    return metalSupported;
+}
+
+JNIEXPORT jboolean JNICALL
+Java_sun_java2d_metal_MTLGraphicsConfig_tryLoadMetalLibrary
+    (JNIEnv *env, jclass mtlgc, jint displayID, jstring shadersLibName)
+{
+  jboolean ret = JNI_FALSE;
+  JNI_COCOA_ENTER(env);
+  NSMutableArray * retArray = [NSMutableArray arrayWithCapacity:3];
+  [retArray addObject: [NSNumber numberWithInt: (int)displayID]];
+  [retArray addObject: [NSString stringWithUTF8String: JNU_GetStringPlatformChars(env, shadersLibName, 0)]];
+  if ([NSThread isMainThread]) {
+      [MTLGraphicsConfigUtil _tryLoadMetalLibrary: retArray];
+  } else {
+      [MTLGraphicsConfigUtil performSelectorOnMainThread: @selector(_tryLoadMetalLibrary:) withObject: retArray waitUntilDone: YES];
+  }
+  NSNumber * num = (NSNumber *)[retArray objectAtIndex: 0];
+  ret = (jboolean)[num boolValue];
+  JNI_COCOA_EXIT(env);
+  return ret;
+}
+
+
+/**
+ * Determines whether the MTL pipeline can be used for a given GraphicsConfig
+ * provided its screen number and visual ID.  If the minimum requirements are
+ * met, the native MTLGraphicsConfigInfo structure is initialized for this
+ * GraphicsConfig with the necessary information (pixel format, etc.)
+ * and a pointer to this structure is returned as a jlong.  If
+ * initialization fails at any point, zero is returned, indicating that MTL
+ * cannot be used for this GraphicsConfig (we should fallback on an existing
+ * 2D pipeline).
+ */
+JNIEXPORT jlong JNICALL
+Java_sun_java2d_metal_MTLGraphicsConfig_getMTLConfigInfo
+    (JNIEnv *env, jclass mtlgc, jint displayID, jstring mtlShadersLib)
+{
+  jlong ret = 0L;
+  JNI_COCOA_ENTER(env);
+  NSMutableArray * retArray = [NSMutableArray arrayWithCapacity:3];
+  [retArray addObject: [NSNumber numberWithInt: (int)displayID]];
+  [retArray addObject: [NSString stringWithUTF8String: JNU_GetStringPlatformChars(env, mtlShadersLib, 0)]];
+  if ([NSThread isMainThread]) {
+      [MTLGraphicsConfigUtil _getMTLConfigInfo: retArray];
+  } else {
+      [MTLGraphicsConfigUtil performSelectorOnMainThread: @selector(_getMTLConfigInfo:) withObject: retArray waitUntilDone: YES];
```

Comment:
You can implement this via ThreadUtilities performOnMainThreadWaiting:YES block:^(). Currently, it has too many utility codes.

Reply:
will be done under JDK-8261143